src/main/java/com/esri/samples/na/routing_around_barriers/RoutingAroundBarriersController.java

```diff
@@ -0,0 +1,331 @@
+        barriersList.add(barrier);
+
+        // build graphics for the barrier and add it to the graphics overlay
+        Graphic barrierGraphic = new Graphic(bufferedBarrierPolygon, barrierSymbol);
+        barriersGraphicsOverlay.getGraphics().add(barrierGraphic);
+      }
+
+    // if the secondary mouse button was clicked, delete the last stop or barrier, respectively
+    } else if (e.getButton() == MouseButton.SECONDARY && e.isStillSincePress()) {
+
+      // check if we can remove stops
+      if (btnAddStop.isSelected() && !stopsList.isEmpty()) {
+        // clear the displayed route, if it exists, since it might not be up to date any more
+        routeGraphicsOverlay.getGraphics().clear();
+
+        // remove the last stop from the stop list and the graphics overlay
+        stopsList.removeLast();
+        stopsGraphicsOverlay.getGraphics().remove(stopsGraphicsOverlay.getGraphics().size() - 1);
+
+      // check if we can remove barriers
+      } else if (btnAddBarrier.isSelected() && !barriersList.isEmpty()) {
+        // clear the displayed route, if it exists, since it might not be up to date any more
+        routeGraphicsOverlay.getGraphics().clear();
+
+        // remove the last barrier from the barrier list and the graphics overlay
+        barriersList.removeLast();
+        barriersGraphicsOverlay.getGraphics().remove(barriersGraphicsOverlay.getGraphics().size() - 1);
+      }
+    }
+  }
+
+  /**
+   * Uses the route task with the lists of stops and barriers to determine the route and display it.
+   */
+  @FXML
+  private void createRouteAndDisplay() {
+    if (stopsList.size() >= 2) {
+      // clear the previous route from the graphics overlay, if it exists
+      routeGraphicsOverlay.getGraphics().clear();
+
+      // add the existing stops and barriers to the route parameters
+      routeParameters.setStops(stopsList);
+      routeParameters.setPolygonBarriers(barriersList);
+
+      // apply the requested route finding parameters
+      routeParameters.setFindBestSequence(findBestSequenceCheckBox.isSelected());
+      routeParameters.setPreserveFirstStop(preserveFirstStopCheckBox.isSelected());
+      routeParameters.setPreserveLastStop(preserveLastStopCheckBox.isSelected());
+
+      // solve the route task
+      final ListenableFuture<RouteResult> routeResultFuture = routeTask.solveRouteAsync(routeParameters);
+      routeResultFuture.addDoneListener(() -> {
+        try {
+          RouteResult routeResult = routeResultFuture.get();
+          if (!routeResult.getRoutes().isEmpty()) {
+            // get the first route result
+            Route firstRoute = routeResult.getRoutes().get(0);
+
+            // create a geometry for this route
+            Geometry routeGeometry = firstRoute.getRouteGeometry();
+
+            // create a graphic for the route and add it to the graphics overlay
+            Graphic routeGraphic = new Graphic(routeGeometry, routeLineSymbol);
```

Comment:
Could probably use a renderer since all routes have the same symbol.

Reply:
That's a great suggestion - should we implement this for Java straight away, or incorporate it into a suggested design enhancement (as discussed below)?